internal/controller/apiextensions/composite/composition_ptf.go

```diff
@@ -271,71 +236,263 @@ func NewPTFComposer(kube client.Client, o ...PTFComposerOption) *PTFComposer {
+	// resource templates. We don't need to do that because we don't support
+	// anonymous resource templates, which can only be associated to existing
+	// composed resources by matching their index in the XR's spec.resourceRefs
+	// array. It's also a lot harder for us to guarantee stable order - our
+	// Functions would have to guarantee stable order too.
+	for i := range cts {
+		ct := cts[i]
+		name := ResourceName(*ct.Name)
+		cd := composed.New()
+
+		// If this template corresponds to an existing observed resource we want
+		// to keep them associated. We copy only the namespace and name, not the
+		// entire observed state, because we're trying to produce only a partial
+		// 'overlay' of desired state.
+		if or, ok := observed[name]; ok {
+			cd.SetNamespace(or.Resource.GetNamespace())
+			cd.SetName(or.Resource.GetName())
+
+			// We want to patch _to_ the XR from observed composed resources,
+			// not from desired state that we've accumulated but not yet
+			// applied. This is because folks will typically be patching from a
+			// field that is set once the observed resource is applied such as
+			// its status. Failures to patch the XR are terminal. We don't want
+			// to apply the XR if a Required patch did not work, for example.
+			if err := RenderToCompositePatches(xr, or.Resource, ct.Patches); err != nil {
+				return CompositionResult{}, errors.Wrapf(err, errFmtRenderToCompositePatches, name)
+			}
+		}
+
+		// Load the P&T base resource template into our empty composed resource.
+		// We consider this a terminal error, since it indicates a broken
+		// CompositionRevision that will never be valid.
+		if err := RenderFromJSON(cd, ct.Base.Raw); err != nil {
+			return CompositionResult{}, errors.Wrapf(err, errFmtParseBase, name)
+		}
+
+		// Failures to patch a composed resource from the environment or from
+		// the XR aren't terminal. We just emit a warning event and move on.
+		// This is because patches often fail because other patches need to
+		// happen first in order for them to succeed. If we returned an error
+		// when a patch failed we might never reach the patch that would unblock
+		// it. We do need to record that the patch failed though, because we
+		// don't want to actually apply the resource. For example we shouldn't
+		// apply a composed resource if a Required patch failed.
+		if err := RenderFromEnvironmentPatches(cd, req.Environment, ct.Patches); err != nil {
+			// Most likely a required FromEnvironment patch failed.
+			events = append(events, event.Warning(reasonCompose, errors.Wrapf(err, errFmtRenderFromCompositePatches, name)))
+			failed[name] = ComposedResourceState{Resource: cd}
+			continue
+		}
+		if err := RenderFromCompositePatches(cd, xr, ct.Patches); err != nil {
+			// Most likely a required FromComposite patch failed.
+			events = append(events, event.Warning(reasonCompose, errors.Wrapf(err, errFmtRenderFromEnvironmentPatches, name)))
```

Comment:
This is the wrong error constant. 😄

Reply:
I'm about to add a commit that addresses https://github.com/crossplane/crossplane/issues/4558 which will make this irrelevant.